### PR TITLE
Validate private key matches a certificate

### PR DESCRIPTION
**Pull request by GreenRover**
We have hat 3 support cases in 3 month where user used private keys that does not related to the used certificate.

The error log from the broker is only relating to not matching modules. 
And our users dont know ssl well / never herd about a "modulus" in ssl context.

### CODE REVIEW

File: src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java [hunk 1]
```diff
@@ -34,14 +37,33 @@ KeyStore createClientKeyStore(final String privateKeyPem, final String clientCer
...
             return keyStore;
         } catch (Exception exception) {
             LOG.warn("Error during keystore creation", exception);
             return null;
         }
     }
 
+    private boolean privateKeyMatchesCertificate(RSAPrivateCrtKey privateKey, Certificate[] certificates) {
+        for (Certificate certificate : certificates) {
+            RSAPublicKey publicKey = (RSAPublicKey) certificate.getPublicKey();
```
Comment (marchof) [inline src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java]: The check works for RSA keys only. Java supports DSA, RSA and EC. Please find a generic test tool (with embedded tests) here: https://gist.github.com/marchof/4e62893b636837411994f6fcf6fc371c

Comment (GreenRover) [reply src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java]: Thanks for code sniped.

File: src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyPairChecker.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            return false;
+        }
+
+        try {
+            var signature = Signature.getInstance(SIGNATURE_ALGORITHMS.get(privateKey.getAlgorithm()));
+            signature.initSign(privateKey);
+            var signed = signature.sign();
+            signature.initVerify(publicKey);
+            return signature.verify(signed);
+        } catch (Exception ignored) {
```
Comment (marchof) [inline src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyPairChecker.java]: If an exception happens in this block there is a serious problem in with our code, as the certificate and the key has already been successfully parsed and constructed. Please catch specific checked exception and log it as ERROR.

File: src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java [hunk 3]
```diff
@@ -34,14 +34,31 @@ KeyStore createClientKeyStore(final String privateKeyPem, final String clientCer
...
+
             return keyStore;
         } catch (Exception exception) {
             LOG.warn("Error during keystore creation", exception);
             return null;
         }
     }
 
+    private boolean privateKeyMatchesCertificate(PrivateKey privateKey, Certificate[] certificates) {
+        for (Certificate certificate : certificates) {
```
Comment (marchof) [inline src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java]: Only the first certificate is the user's certificate where the private key must match. Additional certificates are intermediates where the private key is not used. See [JavaDoc of KeyStore](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/KeyStore.html#getCertificateChain(java.lang.String)):

> the certificate chain (ordered with the user's certificate first followed by zero or more certificate authorities)

File: src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyPairChecker.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        try {
+            var signature = Signature.getInstance(SIGNATURE_ALGORITHMS.get(privateKey.getAlgorithm()));
+            signature.initSign(privateKey);
+            var signed = signature.sign();
+            signature.initVerify(publicKey);
+            return signature.verify(signed);
+        } catch (Exception e) {
+            LOG.error("Unable to load key or cert for matching check", e);
```
Comment (marchof) [inline src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyPairChecker.java]:
```suggestion
            LOG.error("Unable to check key pair", e);
```

The keys are already loaded at this point.

**Review by marchof — APPROVED** [review]

**Pull request merged**